rfcs/0046-versioning-changes-for-hard-fork.md

```diff
@@ -0,0 +1,152 @@
+## Summary
+[summary]: #summary
+
+Remove unneeded version tags from `Bin_prot`-serialized data produced
+by `ppx_version`. Generate version tags on an as-needed basis where
+they're useful. Detect versioning violations in the CI version linter
+in a more useful way.
+
+## Motivation
+[motivation]: #motivation
+
+Versioned data types are built from other versioned types and
+primitive OCaml types. When generating `Bin_prot`-serialized data from
+instances of those type, a version tag is generated not only for the
+containing type, but also for each constituent type. That strategy
+allows determining the version of any piece of the serialized data.
+
+The Jane Street `Async` versioned RPC mechanism used in
+`Mina_networking` already provides versioning of data sent between
+nodes. Therefore, version-tagging of the serialized data is not needed
+for that use case. Removing the version tags will reduce the amount of
+data sent over the network.
+
+We may wish to save serialized data and be able to determine its
+version. For that use case, a single "top" tag for the containing type
+suffices.
+
+Some user-visible data, such as public keys, are formatted using the
+current serialization mechanism, where all contained types have
+version tags. For those few cases, allow use of the existing
+mechanism.
+
+The CI version linter currently compares changes to versioned types
+against the PR base branch. That's more fine-grained than needed, if
+the goal of preventing version changes is to allow interoperability
+with deployed software. Instead, compare the types against the branch 
+of the latest released software. That strategy will reduce the amount
+of noise produced by the linter, and provide more useful results.
+
+## Detailed design
+[detailed-design]: #detailed-design
+
+### Version-tagging changes
+
+Currently, tagging a `Stable` module with `%versioned` produces
+shadowing `bin_read_t` and `bin_write_t` functions that read and write
+version tags, positive integers, prepended to serialized data, via the
+`ppx_version` library. The proposal is not to shadow those functions,
+and instead use those generated by `deriving bin_io`, directly.  It is
+still useful to use `%versioned` to assure that serializations do not
+change for a given versioned type.
+
+For backwards-compatibility, we can add the annotation
+```
+ [@@@with_all_version_tags]
+```
+to a `Stable` module. For each versioned type module `Vn`, generate a
```

Comment:
It seems to me that we don't necessarily want to generate the "all version tags" module for every version into the future, just existing versions that are already out in the wild. As we add new versioned types, I would presume that would want to take advantage of our new top level version tagging system instead. Example given: we need to continue supporting the "all version tags" implementation for existing base58check encoded data, but if we bump the version on any of those, then we could move to a "top version tag" implementation only in the newer versions.

Reply:
We could use the `[@@@with_all_version_tags]` annotation in the `Vn` modules, instead of in the `Stable` module.
Then `With_all_version_tags.bin_read_to_latest` would handle just those legacy versions.

The `[@@@with_top_version_tag]` can remain at the `Stable` level, so we don't have to annotate every `Vn` module.

Reply:
Made this change.